### PR TITLE
Split GitHub actions into different action for each algorithm

### PR DESCRIPTION
**Pull request by MilesCranmer**
This is a draft PR to split the GitHub action by algorithm. Each algorithm will be built and tested in its own separate GitHub action. Since GitHub allows 25 simultaneous actions, this should speed up things greatly.

Draft todo's:

- [ ] Get full suite running correctly.
- [x] Tweak install.sh to only install what is needed for each algorithm.
- [x] Ensure all algorithms are executed.
- [ ] ? (suggestions welcome)

Note that currently PySR's PR is included #62, but these commits will disappear after that one is merged.

### CODE REVIEW

**Comment by MilesCranmer** [comment]
For example, here are all the jobs running on my fork:

Screen Shot 2022-01-21 at 4 01 46 PM

**Comment by MilesCranmer** [comment]
install.sh and environment.yml are now different for each algorithm. This is controlled by, for example, `# only: OperonRegressor` inside the environment.yml file. This seems to work okay

These environments are cached successfully, which reduces build time even more.
Screen Shot 2022-01-22 at 3 18 16 PM

Here's the tests running for each algorithm:
Screen Shot 2022-01-22 at 3 20 59 PM

**Comment by MilesCranmer** [comment]
The file `algorithms.yml` lists the required install files for each algorithm. For example:
```
algorithms:
  - name: AFPRegressor
    install:
      - ellyn_install.sh
```

would mean `ellyn_install.sh` is run for `AFPRegressor`. No other installs are ran because they are not needed for that build.

**Comment by MilesCranmer** [comment]
@lacava can you confirm the contents of `environment.yml` and `algorithms.yml` for me? Thanks.

Also note that this PR is rebased on the PySR one so there's a bunch of other code in this for now. It should be much smaller after the PySR one is through.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ name: CI
 on:
   # Triggers the workflow on push or pull request events but only for the master branch
   push:
-    branches: [ master ]
+    branches: "*"
```
Comment (lacava) [inline .github/workflows/test.yml:9]: why is this changed?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -18,46 +18,78 @@ jobs:
   # This workflow contains a single job called "build"
   build:
     # The type of runner that the job will run on
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        julia-version: ['1.6']
+        os: [ubuntu-latest]
+        algorithm: ["AFPRegressor", "AIFeynman", "BSRRegressor", "DSRRegressor", "EHCRegressor", "FFXRegressor", "FEATRegressor", "FE_AFPRegressor", "EPLEXRegressor", "GPGOMEARegressor", "gplearn", "ITEARegressor", "MRGPRegressor", "OperonRegressor", "PySRRegressor", "sembackpropgp"]
```
Comment (lacava) [inline .github/workflows/test.yml:26]: would be more intelligible to store as a list like 
```
algorithms:
  - name: AFPRegressor
  - name: AIFeynman
  - ...
```

if possible

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -18,46 +18,78 @@ jobs:
...
+          channel-priority: strict
+          # For caching:
+          use-only-tar-bz2: true
+
       - run: |
           conda info
           conda list
           conda activate srbench
 
+      - name: Set up Julia
```
Comment (lacava) [inline .github/workflows/test.yml:69]: not sure about algorithm-specific setups here, eventually should move to docker setup

File: experiment/methods/src/pysr_install.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#install pysr
+
+
+pip install git+git://github.com/MilesCranmer/PySR.git@v0.7.0a2
```
Comment (lacava) [inline experiment/methods/src/pysr_install.sh:4]: rather than pip install here, you can move this to environment.yml following the syntax above.

File: install.sh [hunk 5]
```diff
@@ -10,11 +10,17 @@ echo "///////////////////////////"
 echo "installing GP methods..."
```
Comment (lacava) [inline install.sh:10]: GP should be SR (not your fault, just noticed)

**Comment by lacava** [comment]
First of all, this is great @MilesCranmer , thanks for doing this!

1. Definitely some bugs to work out. I looks like the tests that are "passing" are failing silently. 
2. I see the value in testing in parallel, but I'm not sure about breaking up the environment setup (e.g., installing julia separately). Could lead to unanticipated package conflicts when everything is built together and all methods are tested in the same environment. A cached version of the whole environment would still be fast, I think.

File: .github/workflows/test.yml (hunk 1 above)
Comment (MilesCranmer) [reply .github/workflows/test.yml:9]: This was just for testing on my repo, but should be deleted upon merge.

File: .github/workflows/test.yml (hunk 2 above)
Comment (MilesCranmer) [reply .github/workflows/test.yml:26]: Good idea

**Comment by MilesCranmer** [comment]
Thanks.

The PySR issues are 

> Definitely some bugs to work out. I looks like the tests that are "passing" are failing silently.
Thanks for pointing this out. Maybe I will just leave the environment.yml as-is then, it might be tricky to narrow down which models need what packages.

Indeed it might just be faster to have one single job as is done currently. I'll close this for now.

> I see the value in testing in parallel, but I'm not sure about breaking up the environment setup (e.g., installing julia separately). Could lead to unanticipated package conflicts when everything is built together and all methods are tested in the same environment. A cached version of the whole environment would still be fast, I think.

Let's follow up about this in the PySR pull request #62. I am currently trying to add PySR to conda-forge: https://github.com/conda-forge/staged-recipes/pull/17605, but for now we could try using the conda-forge provided julia version, then include PyJulia and PySR as pip installs.
